### PR TITLE
Change refinement issues and some following issues with empty outputs

### PR DESCRIPTION
**Pull request by KateSakharova**
Fixing 
- checkm2 issue
```
[01/30/2024 03:50:49 PM] INFO: Running CheckM2 version 1.0.1
[01/30/2024 03:50:49 PM] INFO: Custom database path provided for predict run. Checking database at uniref100.KO.1.dmnd...
[01/30/2024 03:51:11 PM] INFO: Running quality prediction workflow with 8 threads.
[01/30/2024 03:51:14 PM] INFO: Calling genes in 1 bins with 8 threads:
    Finished processing 1 of 1 (100.00%) bins.
[01/30/2024 03:51:16 PM] INFO: Calculating metadata for 1 bins with 8 threads:
    Finished processing 1 of 1 (100.00%) bin metadata.
[01/30/2024 03:51:16 PM] INFO: Annotating input genomes with DIAMOND using 8 threads
[01/30/2024 03:52:01 PM] INFO: Processing DIAMOND output
[01/30/2024 03:52:01 PM] ERROR: No DIAMOND annotation was generated. Exiting
```
Issue opened https://github.com/chklovski/CheckM2/issues/95
Checkm2 exited with code 1. But in that case it should not crash. In order to make it continue I catch DIAMOND log in checkm2.log. If it is diamond - checkm2 returns emty files, otherwise it should fail
- support less then 3 binners for prok refinement. I modified python scripts to process empty folders and 1 or 2 binners. There are many ifEmpty appeared.
- changed eukcc container fixing `no tax issue`
- added fixes to checkm and drep for input folders structure
- added final step to finalize logging by runs to track filtering process by steps
example
```
SRR4831653
GGP:EUK_MAGS_GENERATION:EUKCC_CONCOCT:bins: 8, merged: 0
GGP:PROK_MAGS_GENERATION:BIN_REFINEMENT:RENAME_AND_CHECK_SIZE_BINS_BINNER1:bins_dir: 8, filtered_out: 2
GGP:PROK_MAGS_GENERATION:BIN_REFINEMENT:RENAME_AND_CHECK_SIZE_BINS_BINNER2:bins_dir: 2, filtered_out: 2
GGP:PROK_MAGS_GENERATION:BIN_REFINEMENT:RENAME_AND_CHECK_SIZE_BINS_BINNER3:bins_dir: 0, filtered_out: 0
GGP:PROK_MAGS_GENERATION:BIN_REFINEMENT:CHECKM2_BINNER1:filtered: 0
GGP:PROK_MAGS_GENERATION:BIN_REFINEMENT:CHECKM2_BINNER2:filtered: 0
```
- changed QS50 for euks to completeness > 50 & contamination < 5
- renamed eukcc merged bins
- added concatenated depth file for euks coverage recycler

### CODE REVIEW

File: modules/local/align_bwa/main.nf [hunk 1]
```diff
@@ -27,7 +27,8 @@ process INDEX_FASTA {
 
 process ALIGNMENT {
 
-    label 'process_medium'
+    errorStrategy = { task.exitStatus in ((130..145) + 104) ? 'retry' : 'finish' }
```
Comment (mberacochea) [inline modules/local/align_bwa/main.nf:30]: Shouldn't this be in the modules.config file?

File: modules/local/checkm2/main.nf [hunk 2]
```diff
@@ -1,13 +1,18 @@
 process CHECKM2 {
 
     label 'process_medium'
+    errorStrategy =
```
Comment (mberacochea) [inline modules/local/checkm2/main.nf:4]: Shouldn't this be in the modules.config file?

File: modules/local/checkm2/main.nf [hunk 3]
```diff
@@ -18,12 +23,40 @@ process CHECKM2 {
...
         --database_path ${checkm2_db}
+    CHECMK2_EXITCODE="\$?"
+
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        checkm2: \$(checkm2 --version)
+    END_VERSIONS
+
+    if [ "\$CHECMK2_EXITCODE" == "1" ]; then
+        echo "Checkm2 exit code \$CHECMK2_EXITCODE"
```
Comment (mberacochea) [inline modules/local/checkm2/main.nf:42]: Should we print these in the .error file?

`echo "Error" >&2`

File: modules/local/checkm2/main.nf [hunk 4]
```diff
@@ -18,12 +23,40 @@ process CHECKM2 {
...
+            exit \$CHECMK2_EXITCODE
+        else
+            echo "checkm2.log exists -> checking for diamond error"
+            if grep -q "No DIAMOND annotation was generated. Exiting" "${name}_checkm_output/checkm2.log"; then
+                echo "No DIAMOND annotation was generated"
+                touch "${name}_all_stats.csv" "${name}_filtered_genomes.tsv"
+                mkdir "${name}_filtered_genomes"
+                exit 0
+            else
+                echo "It is not Diamond, sorry, check manually. Exit"
```
Comment (mberacochea) [inline modules/local/checkm2/main.nf:54]:
```suggestion
                echo "It is not DIAMOND, sorry, check manually. Exit" >&2
```

File: modules/local/checkm2/main.nf [hunk 5]
```diff
@@ -18,12 +23,40 @@ process CHECKM2 {
...
+        checkm2: \$(checkm2 --version)
+    END_VERSIONS
+
+    if [ "\$CHECMK2_EXITCODE" == "1" ]; then
+        echo "Checkm2 exit code \$CHECMK2_EXITCODE"
+        if [ ! -e "${name}_checkm_output/checkm2.log" ]; then
+            echo "checkm2.log does not exist. Exit"
+            exit \$CHECMK2_EXITCODE
+        else
+            echo "checkm2.log exists -> checking for diamond error"
```
Comment (mberacochea) [inline modules/local/checkm2/main.nf:47]:
```suggestion
            echo "checkm2.log exists -> checking if DIAMOND failed"
```

**Review by mberacochea — COMMENTED** [review]
:+1: excellent.. I left some very minor comments

File: modules/local/checkm2/main.nf (hunk 2 above)
Comment (KateSakharova) [reply modules/local/checkm2/main.nf:4]: I can add it there but I was not sure will nextflow take this errorStrategy or errorStrategy from process_medium from base.config...

Comment (KateSakharova) [reply modules/local/checkm2/main.nf:4]: actually, those errorStrategies are the same as in base.config. I will remove them from modules at all :)

**Comment by KateSakharova** [comment]
I will close that PR because it is getting massive. It fixed known issues 🤞

**Pull request merged**